cirq/protocols/qid_shape_protocol.py

```diff
@@ -0,0 +1,168 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Any, Tuple, TypeVar, Union
+
+from typing_extensions import Protocol
+
+from cirq.type_workarounds import NotImplementedType
+
+if TYPE_CHECKING:
+    # pylint: disable=unused-import
+    import cirq
+
+# This is a special indicator value used by the methods to determine whether or
+# not the caller provided a 'default' argument. It must be of type
+# Tuple[int, ...] to ensure the method has the correct type signature in that
+# case. It is checked for using `is`, so it won't have a false positive if the
+# user provides a different (0,) value.
+RaiseTypeErrorIfNotProvided = (0,)  # type: Any
+# Equal integers outside the range [-5, 256] aren't identically equal with `is`.
+RaiseTypeErrorIfNotProvidedInt = -2**512  # type: Any
+
+TDefault = TypeVar('TDefault')
+
+
+class SupportsQidShape(Protocol):
+    """A unitary, channel, mixture or other object that operates on a known
+    number qubits/qudits/qids, each with a specific number of quantum levels."""
+
+    def _qid_shape_(self) -> Union[Tuple[int, ...], NotImplementedType]:
+        """A tuple specifying the number of quantum levels of each qid this
+        object operates on, e.g. (2, 2, 2) for a three-qubit gate.
+
+        This method is used by the global `cirq.qid_shape` method (and by
+        `cirq.num_qubits` if `_num_qubits_` is not defined). If this
+        method is not present, or returns NotImplemented, it is assumed that the
+        receiving object operates on qubits. (The ability to return
+        NotImplemented is useful when a class cannot know if it has a shape
+        until runtime.)
+
+        The order of values in the tuple is always implicit with respect to the
+        object being called. For example, for gates the tuple must be ordered
+        with respect to the list of qubits that the gate is applied to. For
+        operations, the tuple is ordered to match the list returned by its
+        `qubits` attribute.
+
+        Returns:
+            A unitary matrix describing this value, or NotImplemented if the
+            shape is unknown.
+        """
+
+    def _num_qubits_(self) -> Union[int, NotImplementedType]:
+        """The number of qubits this object operates on.
+
+        This method is used by the global `cirq.num_qubits` method (and by
+        `cirq.qid_shape` if `_qid_shape_` is not defined.  If this
+        method is not present, or returns NotImplemented, it will fallback
+        to using the length of `_qid_shape_`.
+
+        Returns:
+            An integer specifying the number of qubits.
+        """
+
+
+def qid_shape(val: Any, default: TDefault = RaiseTypeErrorIfNotProvided
```

Comment:
We should do something reasonable for "circuit-esque" values such as Iterable[Operation]. If the input doesn't have an explicit method, but is iterable and only contains operations (perhaps we also need a qubits protocol to avoid the type specificity), then the `qubits` and `qid_shape` methods should do the thing you'd expect `Circuit` to do.

Reply:
I implemented `cirq.max_qid_shape(op_tree)` to replace `Circuit.max_qid_shape()`.